com.playeveryware.eos/Runtime/Core/EOSManager.cs

```diff
@@ -1891,5 +1903,46 @@ void IEOSCoroutineOwner.StartCoroutine(IEnumerator routine)
         {
             base.StartCoroutine(routine);
         }
+
+        /// <summary>
+        /// Enqueues an Action to be executed on the main thread.
+        /// </summary>
+        /// <param name="action">Action to execute.</param>
+        static public void DispatchAsync(Action action)
+        {
+            lock (s_enqueuedTasksLock)
+            {
+                // Lazy allocate the queue
+                if (s_enqueuedTasks == null)
+                {
+                    s_enqueuedTasks = new List<Action>();
+                }
+                s_enqueuedTasks.Add(action);
+            }
+        }
+
+        static private void ExecuteQueuedMainThreadTasks()
+        {
+            // Lock the enqued tasks list, and hold refrence to the enqueued tasks.
+            // This is done so that the foreach loop doesn't potentially go "forever"
+            // if a given action in the queue happens to generate a list of tasks that 
+            // also generate a list of task. The s_enqueuedTasks list is also nulled out
+            // because we only allocate the list if we need to queue up new tasks. See DispatchSync
+            List<Action> actionsToRun;
+            lock (s_enqueuedTasksLock)
+            {
+                actionsToRun = s_enqueuedTasks;
+                s_enqueuedTasks = null;
+                if (actionsToRun == null)
+                {
+                    return;
+                }
+            }
+
+            foreach (Action action in actionsToRun)
+            {
+                action.Invoke();
+            }
```

Comment:
I _could_ be wrong about this - but don't we want the lock to enclose the invocation of the actions as well?

Reply:
This was copied from elsewhere, but looking at it... Yea I can't see why the lock is outside of the actual execution.

Reply:
Ah, I see now. This is outside of it in case anything wants to run `DispatchAsync` during something that is executing in `ExecuteQueuedMainThreadTasks`. If there was a lock around the whole thing, then there'd be a deadlock when executing the tasks. Instead;

- Create `actionsToRun`
- Lock on the `s_enqueuedTasksLock`
- Set `actionsToRun` to the current `s_enqueuedTasks`
- End lock
- If there was anything in `actionsToRun`, then execute them

I think this is the correct implementation, as-is.